### PR TITLE
DEVOPS-452: Update package with python-conda-template

### PR DESCRIPTION
**Pull request by SophieCurinier**
**DEVOPS-452 - apply latest version of python template repos to all repos**
- Use reusable github workflows
- Use MiraGeoscience/pre-commit-hooks package in `.pre-commit-hooks.yml`
- Update pre-commit-hooks version:
   - `ruff-pre-commit` : v0.5.1
   - `mirrors-mypy`: v1.10.1
   - `Mirageoscience/pre-commit-hooks`: v1.0.1
   - `codespell`: v2.3.0
   - `rstcheck`: v6.2.4

### CODE REVIEW

**Comment by SophieCurinier** [comment]
There is issue in the code due to the dependency octree-creation-app, mira-simpeg and geoapps-utils. This issue need to be check by the Python team.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -26,21 +21,28 @@ repos:
 #        -   --unsafe-fixes
     -   id: ruff-format
 -   repo: https://github.com/pre-commit/mirrors-mypy
-    rev: v1.6.1
+    rev: v1.11.0
     hooks:
     -   id: mypy
         additional_dependencies: [
-            numpy==1.26.*,  # pre-commit CI does not provide numpy 1.23.*
+            numpy==1.23.*,
```
Comment (sebhmg) [inline .pre-commit-config.yaml:28]: now using 1.26 (also pre-commit CI does not support numpy 1.23)
```suggestion
            numpy==1.26.*,
```

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -26,21 +21,28 @@ repos:
...
             types-toml,
             types-PyYAML,
-            tomli, # to read config from pyproject.toml
+            tomli # to read config from pyproject.toml
         ]
```
Comment (sebhmg) [inline .pre-commit-config.yaml:32]:
```suggestion
            tomli # to read config from pyproject.toml
        ]
       exclude: (conf\.py|tests)
```

File: docs/source/conf.py [hunk 3]
```diff
@@ -14,10 +14,10 @@
 # -- General configuration ---------------------------------------------------
 # https://www.sphinx-doc.org/en/master/usage/configuration.html#general-configuration
 
-extensions = []
+extensions = []  # type: list[str]
```
Comment (sebhmg) [inline docs/source/conf.py:17]: mypy not supposed to run here, thanks to exclusion pattern in pre-commit config
```suggestion
```

File: docs/source/conf.py [hunk 4]
```diff
@@ -14,10 +14,10 @@
...
 
 templates_path = ["_templates"]
-exclude_patterns = []
+exclude_patterns = []  # type: list[str]
```
Comment (sebhmg) [inline docs/source/conf.py:20]: mypy not supposed to run here, thanks to exclusion pattern in pre-commit config
```suggestion
exclude_patterns = []
```

**Review by sebhmg — APPROVED** [review]

**Review by sebhmg — CHANGES_REQUESTED** [review]
jobs are now failing but they used to pass on the develop branch

**Comment by SophieCurinier** [comment]
> jobs are now failing but they used to pass on the develop branch.
> 
> pyproject.toml of geoapps-utils it is pointing at seems invalid: ` ERROR: Could not build wheels for geoapps-utils, which is required to install pyproject.toml-based projects`

I'm not sure how to do it. The issue come from the imported packages. 😞

**Comment by sebhmg** [comment]
merged through https://github.com/MiraGeoscience/plate-simulation/pull/30

**Pull request closed without merging**